File: oran_drishti_final/model_engine.py
"""
model_engine.py — Trishul AI Core: Three Independent Filters
=============================================================
Implements the three scientific filters that together form the
Oran-Drishti anomaly detection pipeline.  All three must agree
before an "Adharma Alert" (illegal mining alert) is raised.

Filter 1 — Ritu-Chakra    : FFT seasonal decomposition
Filter 2 — Shuddhi Model  : LSTM autoencoder anomaly scoring
Filter 3 — Surya-Sakshi   : Solar-physics shadow depth check
"""

import numpy as np
from scipy.fft import fft, fftfreq
import datetime
from pysolar.solar import get_altitude
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, RepeatVector, TimeDistributed, Dense


# ============================================================
# FILTER 1: Ritu-Chakra — Harmonic Analysis via FFT
# ============================================================
def extract_harmonics(time_series, sample_spacing=1.0):
    """
    Prong 1: Ritu-Chakra (Harmonic Analysis)

    Uses Fast Fourier Transform (FFT) to decompose the time-series
    into its constituent frequencies.  This isolates the natural
    seasonal 'heartbeat' (Ritu / Seasonal Cycle) from chaotic noise
    or a structurally broken flatline caused by mining degradation.

    A healthy forest exhibits a strong annual/semi-annual harmonic.
    A mined plot shows a near-zero amplitude across all frequencies
    (flatline in time domain → no dominant seasonal peak in frequency
    domain).

    Args:
        time_series (array-like): Temporal NDVI values (one per month).
        sample_spacing (float):   Time interval between samples in
                                  consistent units (default: 1 month).

    Returns:
        xf (ndarray):         Positive frequency bins (cycles / sample).
        amplitudes (ndarray): Magnitude of each frequency component,
                              normalised so that amplitudes sum to the
                              RMS of the original signal.
    """
    # Ensure the input is a contiguous NumPy array for SciPy FFT
    ts = np.array(time_series)
    N  = len(ts)

    # Compute the full FFT (complex result)
    yf = fft(ts)

    # Retain only the positive half of the spectrum (Nyquist theorem):
    # frequencies range from 0 to (N/2 - 1) / (N * sample_spacing)
    xf = fftfreq(N, sample_spacing)[:N // 2]

    # Normalise amplitudes by 2/N so that the peak height matches
    # the true sinusoidal amplitude in the original signal.
    amplitudes = 2.0 / N * np.abs(yf[0:N // 2])

    return xf, amplitudes


# ============================================================
# FILTER 2: Shuddhi Model — LSTM Autoencoder (Unsupervised)
# ============================================================
def build_lstm_autoencoder(timesteps, features):
    """
    Prong 2: Shuddhi Model (Unsupervised Deep Learning)

    Builds a Sequence-to-Sequence LSTM Autoencoder to learn the
    'Natural Order' (Dharma) of a healthy forest pixel's time-series.

    Architecture (Encoder → Bottleneck → Decoder):
    ┌──────────────────┐
    │  Input           │  shape: (timesteps, features)
    │  LSTM Encoder    │  units=16, compresses to a latent vector
    │  RepeatVector    │  broadcasts latent vector across time steps
    │  LSTM Decoder    │  units=16, reconstructs the sequence
    │  TimeDistributed │  Dense(features) — per-step reconstruction
    └──────────────────┘

    Training strategy (unsupervised / self-supervised):
    - Train ONLY on "healthy" pixels (first 5×5 corner of the spatial
      grid, assumed to be undisturbed baseline forest).
    - The autoencoder learns to reconstruct normal seasonal patterns.
    - At inference, a degraded/mining pixel produces high MSE because
      its flatline pattern was never seen during training.

    Args:
        timesteps (int): Number of time steps per sample (e.g., 36 months).
        features  (int): Feature dimension per step (e.g., 1 for NDVI).

    Returns:
        model (Sequential): Compiled Keras model ready for `.fit()`.
    """
    model = Sequential()

    # ---- Encoder ----
    # LSTM with return_sequences=False collapses the temporal dimension
    # into a single fixed-size latent vector of shape (batch, units).
    model.add(LSTM(
        16,
        activation='relu',
        input_shape=(timesteps, features),
        return_sequences=False  # Compress entire sequence to one vector
    ))

    # ---- Bottleneck ----
    # RepeatVector repeats the latent vector `timesteps` times to give
    # the decoder one copy per time step: shape → (batch, timesteps, 16)
    model.add(RepeatVector(timesteps))

    # ---- Decoder ----
    # LSTM with return_sequences=True outputs a hidden state at each step,
    # enabling per-step reconstruction: shape → (batch, timesteps, 16)
    model.add(LSTM(16, activation='relu', return_sequences=True))

    # ---- Output ----
    # TimeDistributed wraps Dense so the same linear layer is applied
    # independently to each time step: shape → (batch, timesteps, features)
    model.add(TimeDistributed(Dense(features)))

    # MSE loss = reconstruction error; Adam converges quickly for demo
    model.compile(optimizer='adam', loss='mse')
    return model


# ============================================================
# FILTER 3: Surya-Sakshi — Solar Physics Validator
# ============================================================
def calculate_sun_angle(lat, lon, date=None):
    """
    Prong 3: Surya-Sakshi — Shadow Depth Validation

    Calculates the solar altitude angle (degrees above the horizon)
    for a given geographic location and UTC datetime using the
    high-precision `pysolar` library.

    Purpose:
    If the sun is high (> 30°) at the time of image acquisition and
    the model still detects a deep dark anomaly, the shadow cannot
    originate from a hillside or a low-angle illumination artefact.
    It must be caused by volumetric depth — i.e., an open-cast
    mining pit.  This is the Surya-Sakshi (Sun Witness) validation.

    Args:
        lat  (float):            Target latitude in decimal degrees.
        lon  (float):            Target longitude in decimal degrees.
        date (datetime, optional): UTC-aware datetime of acquisition.
                                   Defaults to 2025-05-15 12:00 UTC
                                   (representative Aravalli dry-season
                                   acquisition timestamp).

    Returns:
        altitude (float): Solar elevation angle in degrees.
                          > 0  → daytime;  ≤ 0 → night / below horizon.
    """
    # Default to a fixed demo timestamp if no date is provided
    if date is None:
        date = datetime.datetime(2025, 5, 15, 12, 0, 0,
                                 tzinfo=datetime.timezone.utc)

    # pysolar.get_altitude() returns degrees; positive = above horizon
    return get_altitude(lat, lon, date)

File: src/utils/modelEngine.ts
// ============================================================
// modelEngine.ts — Oran-Drishti Three-Filter ML Pipeline
//
// Implements the three independent validation filters that
// together form the "Trishul AI" anomaly detection system:
//
//   Filter 1 — Ritu-Chakra   : FFT harmonic analysis (seasonality)
//   Filter 2 — Shuddhi Model : LSTM autoencoder (anomaly scoring)
//   Filter 3 — Surya-Sakshi  : Solar physics (shadow depth check)
//
// All three filters must agree before an "Adharma Alert" is raised,
// greatly reducing false positives from natural phenology changes.
// ============================================================

import SunCalc from 'suncalc';
import { TimeSeriesPoint } from './dataSimulator';
import * as tf from '@tensorflow/tfjs';
import FFT from 'fft.js';

// ============================================================
// FILTER 1: Ritu-Chakra — FFT Harmonic Analysis
// ============================================================
// Decomposes the NDVI time-series into its constituent frequencies
// using a Fast Fourier Transform, then reconstructs only the
// low-frequency (seasonal) components.  The returned "seasonal"
// field on each point represents the expected natural cycle;
// anything the model cannot explain with those harmonics is a
// candidate anomaly for Filter 2 to evaluate.
export const performHarmonicAnalysis = (history: TimeSeriesPoint[]): TimeSeriesPoint[] => {
  const n = history.length;

  // FFT.js requires the input length to be an exact power of 2.
  // We find the next power of 2 >= n and zero-pad the signal.
  const nextPow2 = Math.max(2, Math.pow(2, Math.ceil(Math.log2(n))));
  
  // Allocate interleaved real/imaginary arrays (fft.js convention:
  // even indices = real parts, odd indices = imaginary parts).
  const f = new FFT(nextPow2);
  const input = f.createComplexArray();   // Will hold the time-domain signal
  const output = f.createComplexArray();  // Will hold the frequency-domain result

  // Populate the real component with NDVI values; imaginary part stays 0.
  for (let i = 0; i < nextPow2; i++) {
    if (i < n) {
      input[2 * i]     = history[i].ndvi; // Real part: actual NDVI value
      input[2 * i + 1] = 0;               // Imaginary part: zero (real signal)
    } else {
      // Zero-padding: extends the array to the next power-of-2 length
      input[2 * i]     = 0;
      input[2 * i + 1] = 0;
    }
  }

  // --- Forward FFT: time domain → frequency domain ---
  f.transform(output, input);

  // --- Low-pass filter: zero out all high-frequency bins ---
  // Bins 0..cutoff correspond to the DC offset and the lowest annual/
  // bi-annual harmonics that capture natural phenology.
  // Bins beyond `cutoff` (and their mirrored counterparts) are noise or
  // sudden structural changes — we remove them to reconstruct only the
  // expected seasonal baseline.
  const cutoff = 10; // Keep DC + first 10 harmonics
  for (let i = cutoff; i < nextPow2 - cutoff; i++) {
    output[2 * i]     = 0; // Zero real component of high-frequency bin
    output[2 * i + 1] = 0; // Zero imaginary component of high-frequency bin
  }

  // --- Inverse FFT: filtered frequency domain → smoothed time domain ---
  f.inverseTransform(input, output);

  // fft.js inverse transform is unscaled, so divide by the FFT size to
  // recover values in the original NDVI magnitude range (roughly 0–1).
  const smoothed = history.map((point, i) => {
    return { ...point, seasonal: input[2 * i] / nextPow2 }; 
  });

  return smoothed; // Each point now has a `seasonal` field = expected NDVI
};

// ============================================================
// FILTER 2: Shuddhi Model — LSTM Autoencoder (TensorFlow.js)
// ============================================================
// Trains an LSTM autoencoder entirely in the browser on the first
// 50 % of the NDVI time-series (the "healthy" baseline period).
// After training, it tries to reconstruct the full series.
// Pixels/timesteps where reconstruction error exceeds the threshold
// are flagged as anomalies — they represent patterns the model has
// "never seen" during healthy training, implying structural change.
//
// This function is async because TensorFlow.js model training is
// asynchronous (WebGL/WASM kernels).
export const trainAndPredictAnomalies = async (history: TimeSeriesPoint[], onProgress?: (epoch: number, loss: number) => void) => {
  
  // --- Step 1: Build sliding-window input tensors ---
  // Each training sample is a sequence of `windowSize` consecutive NDVI
  // readings.  We slide this window one step at a time across the series.
  const windowSize = 30; // 30-day context window
  const inputs: number[][] = [];
  
  const values = history.map(p => p.ndvi); // Extract raw NDVI numbers
  
  for (let i = 0; i < values.length - windowSize; i++) {
    inputs.push(values.slice(i, i + windowSize)); // One window per sample
  }

  // Shape: [numWindows, windowSize, 1] — last dim is the single NDVI feature
  const xs = tf.tensor3d(inputs, [inputs.length, windowSize, 1]);
  
  // Use only the first 50 % of windows as the "healthy" training set.
  // This assumes the earlier part of the time-series was undisturbed.
  const trainSize = Math.floor(inputs.length * 0.5);
  const trainXs = xs.slice([0, 0, 0], [trainSize, windowSize, 1]);

  // --- Step 2: Define the LSTM Autoencoder architecture ---
  // Encoder compresses the sequence to a latent vector;
  // Decoder reconstructs the sequence from that vector.
  const model = tf.sequential();
  
  // Encoder: single LSTM layer that returns only the final hidden state
  // (returnSequences: false), producing a fixed-size latent vector.
  model.add(tf.layers.lstm({
    units: 32,
    inputShape: [windowSize, 1],
    returnSequences: false // Output shape: [batch, 32]
  }));
  
  // Bottleneck: repeat the latent vector `windowSize` times so the
  // decoder receives one copy per time step.
  model.add(tf.layers.repeatVector({ n: windowSize })); // Output: [batch, windowSize, 32]
  
  // Decoder: LSTM that reconstructs the sequence step by step.
  model.add(tf.layers.lstm({
    units: 32,
    returnSequences: true // Output shape: [batch, windowSize, 32]
  }));
  
  // Project decoder output back to a single feature (NDVI) per step.
  model.add(tf.layers.timeDistributed({ layer: tf.layers.dense({ units: 1 }) }));

  model.compile({ optimizer: 'adam', loss: 'meanSquaredError' });

  // --- Step 3: Train on the healthy baseline (first 50 % of data) ---
  await model.fit(trainXs, trainXs, {  // Autoencoder target = input
    epochs: 15,    // 15 epochs is fast enough for a browser demo
    batchSize: 32,
    shuffle: true,
    callbacks: {
      // Report per-epoch loss to the UI so users can see training progress
      onEpochEnd: (epoch, logs) => {
        if (onProgress && logs) onProgress(epoch, logs.loss);
      }
    }
  });

  // --- Step 4: Reconstruct ALL windows (not just the training half) ---
  // The model will have low error on healthy data and high error on the
  // anomalous (degraded) portion it was never trained on.
  const preds = model.predict(xs) as tf.Tensor;
  const reconstructed = await preds.array() as number[][][]; // [numWindows, windowSize, 1]
  
  // --- Step 5: Score each time-series point by reconstruction error ---
  const anomalies = history.map((point, i) => {
    // The first `windowSize` points have no complete preceding window,
    // so we cannot score them — mark as non-anomalous.
    if (i < windowSize) return { ...point, reconstructionError: 0, isAnomaly: false };
    
    const windowIndex = i - windowSize;
    if (windowIndex >= reconstructed.length) return { ...point, reconstructionError: 0, isAnomaly: false };

    // Compare the reconstruction of the LAST step in the sliding window
    // (which corresponds to time step `i`) against the true observed value.
    const predVal  = reconstructed[windowIndex][windowSize - 1][0];
    const actualVal = values[i];
    const error = Math.abs(predVal - actualVal); // Mean Absolute Error

    return {
      ...point,
      reconstructionError: error,
      isAnomaly: error > 0.15 // Anomaly threshold (tunable; 0.15 ~ 15 % NDVI deviation)
    };
  });

  // --- Step 6: Dispose all TF.js tensors to free GPU/WASM memory ---
  xs.dispose();
  trainXs.dispose();
  preds.dispose();
  model.dispose();

  return anomalies;
};

// ============================================================
// FILTER 3: Surya-Sakshi — Solar Physics Validator
// ============================================================
// Uses the SunCalc library to compute the precise solar altitude
// and azimuth for a given location and timestamp.  Shadow length
// is derived from the sun's elevation angle.
//
// This filter is used as a final confirmation step: if the sun is
// high in the sky (> 30°) and the anomaly score is still elevated,
// the dark region is likely volumetric (i.e., a mining pit casting
// a deep shadow) rather than a surface artefact like a cloud shadow.
export const validateSolarShadows = (lat: number, lon: number, dateStr: string) => {
  const date = new Date(dateStr);

  // SunCalc.getPosition returns altitude and azimuth in radians;
  // we convert to degrees for human-readable display.
  const sunPos = SunCalc.getPosition(date, lat, lon);
  const altitude = sunPos.altitude * (180 / Math.PI); // Elevation above horizon (°)
  const azimuth  = sunPos.azimuth  * (180 / Math.PI); // Compass bearing (°)
  
  return {
    altitude: altitude.toFixed(2),
    azimuth:  azimuth.toFixed(2),
    isDaytime: altitude > 0, // Sun is above the horizon
    // Shadow length relative to object height: length = 1 / tan(altitude)
    // Returns 'Infinite' at night (altitude ≤ 0) to indicate no shadow analysis.
    shadowLengthFactor: altitude > 0 ? (1 / Math.tan(sunPos.altitude)).toFixed(2) : 'Infinite'
  };
};
